### PR TITLE
graph-tool: 2.18

### PR DESCRIPTION
**Pull request by count0**

### CODE REVIEW

File: graph-tool.rb [hunk 1]
```diff
@@ -30,12 +11,6 @@ class GraphTool < Formula
     depends_on "libtool" => :build
   end
 
-  bottle do
-    sha256 "1403a9699ab147da134b39d4cc3b64cd1a49be1d7d9b5941a842f87f5b9132b6" => :el_capitan
-    sha256 "c0259303a55befb147a40d60e58e283d2b9774d5a52b9d7f13b8eacc8e0d063f" => :yosemite
-    sha256 "7f2ecc68e838a493bcfe533b36c94d1e2b7dc90bd03ea0e1453f94bc1d80ef11" => :mavericks
-  end
-
```
Comment (dpo) [inline graph-tool.rb]: Please don't alter the `bottle` section.

**Comment by dpo** [comment]
```
Error: 3 problems in 1 formula
homebrew/science/graph-tool:
 * revision should be removed
 * 13: Trailing whitespace was found
 * 19: Trailing whitespace was found
```
Please build and audit locally before submitting.

**Comment by count0** [comment]
@dpo can't really build locally, sorry.

**Comment by count0** [comment]
@dpo One of the test failures (on 10.10) was due to a timeout. Would it be possible to restart it? The failure on 10.9 ("dyld: lazy symbol binding failed") seems like a compiler/linker bug, since the symbols in question need to be generated automatically by the compiler. Is there a mechanism to flag/exclude a given OS version?

**Comment by dpo** [comment]
Doesn't look like an appropriate compiler was selected.

**Comment by count0** [comment]
@dpo This is a bit of whack-a-mole... The only compilers that will work are clang >= 3.7.1 and gcc >=5, but here we have to rule out all others one by one. Is there a more straightforward way?

**Comment by count0** [comment]
@dpo I'm quite stuck here. I think I might drop this merge request. In macports this works flawlessly, since it easier to select the right compiler. But here we have to flag all that **don't** work, but there is only two that will: the newest vanilla clang and gcc. I don't even know all other possible compiler versions and build numbers out there in homebrew...

**Comment by dpo** [comment]
@count0 Should we simply force gcc using `fails_with :clang` without specifying a build number?

**Comment by count0** [comment]
@dpo GCC should work, but it uses libstdc++ per default. Clang uses libc++. These two don't mix. Hence, if something else (e.g. boost) was compiled with clang/libc++ it will not work with graph-tool compiled with gcc/libstdc++. It could be that homebrew does things right and always selects the same C++ library independent of compiler, but I'm not sure.

**Comment by dpo** [comment]
That works, but how about #3532? It seems to succeed on 10.9 with your previous exclusions.

**Comment by count0** [comment]
@dpo I updated it to the current version too.

**Comment by dpo** [comment]
@count0 Sorry for the back-and-forth but I just found [this](https://trac.macports.org/changeset/146449) linked from one of your issues. How about
```ruby
fails_with :clang do
  build 700.1.81
end
```
but it seems it might not be necessary with version 2.14?!

**Comment by count0** [comment]
AFAIK nothing really changed in version 2.14 with regards to this problem... If the current code does not trigger the bug any longer, it would be by accident. However, I have not seen any evidence of this.

I had not seen this specific build mask from macports.  I will put it in the formula, and we will see how far we go.

**Comment by dpo** [comment]
Sorry, should be `build 700`.

**Comment by dpo** [comment]
ps: why depend on `homebrew/science/numpy` explicitly? Is there something specific about that `numpy`, or could you just say `depends_on "numpy" => :python`?

**Comment by count0** [comment]
I can't see any particular reason (I'm not the original author of this formula). We do need the C API from numpy to be available, but it should be there in any complete numpy installation.

But let us wait and see if the current version works at all. If it does, we can make this other change.

**Comment by dpo** [comment]
Locally, on 10.9, it builds but I get the same lazy symbol error.

**Comment by count0** [comment]
Which version of clang got selected?

**Comment by dpo** [comment]
gcc-5 because I only have clang 600 on this machine.

**Comment by count0** [comment]
Weird, I've never seen this issue on GCC before. Maybe it is related to libc++, not clang itself?

What specific version of GCC 5 do you have?

**Comment by dpo** [comment]
I have gcc 5.3.0.

**Comment by dpo** [comment]
If `boost-python` is built with clang, it uses libc++.

**Comment by count0** [comment]
@dpo OK, then this is most likely a libc++ issue. Can you verify which version do you have installed?

**Comment by dpo** [comment]
I have libc++ 1.0. Is there a leaner version of `graph-tool` we could experiment with? It takes forever to build.

**Comment by dpo** [comment]
I don't get the lazy symbol error when `boost`, `boost-python` and `graph-tool` are all compiled with gcc-5 on OSX 10.9.

**Comment by count0** [comment]
Ok, so that was indeed a compiler issue. It must also work with vanilla clang >= 3.7.

**Comment by dpo** [comment]
Ok, but I'm not sure we can force dependencies to also be built with a certain compiler.

**Comment by count0** [comment]
Actually, since you used GCC for everything, it probably used libstdc++, right? Hence we cannot rule out that the problem is actually with libc++. It is possible that we can have only graph-tool compiled with GCC + libc++, provided libc++ is new enough. However, I'm not even sure homebrew has a separate formula for libc++; it could just come bundled with clang. If that is the case, we are bach at square one with the whole problem of selecting the right clang version...

**Comment by count0** [comment]
So, from what I understand this really affects only 10.9, since clang 3.7 is not available there? Would a reasonable compromise be simply to display a warning for 10.9 that boost and boost_python must be compiled with the same compiler?

**Comment by dpo** [comment]
There's no formula for libc++ as it ships with clang. But we can query how `boost` and `boost-python` were installed. For example:
```
$ brew irb
==> Interactive Homebrew Shell
Example commands available with: brew irb --examples
irb(main):004:0> Tab.for_name("boost").stdlib
=> "libcxx"
irb(main):005:0> Tab.for_name("boost").compiler
=> "clang"
irb(main):007:0> Tab.for_name("boost-python").stdlib
=> "libcxx"
irb(main):008:0> Tab.for_name("boost-python").compiler
=> "clang"
```
On OSX 10.9, the latest Xcode is version 6.2, so clang > 700 is not even available. We can simply fail in that case I guess. Something along the lines of
```ruby
fails_with :clang
  build 700
end

...

if MacOS.version == :mavericks && (Tab.for_name("boost").stdlib == "libcxx" || Tab.for_name("boost-python").stdlib == "libcxx")
  odie "boost and boost-python must be built against libstdc++ on Mavericks"
end
```
I'm not quite sure how the build bot will react though.

**Comment by count0** [comment]
Well, let's try.

File: graph-tool.rb [hunk 2]
```diff
@@ -66,6 +66,10 @@ class GraphTool < Formula
     cause "We need GCC 5.0 or above for sufficient c++14 support"
   end
 
+  if MacOS.version == :mavericks && (Tab.for_name("boost").stdlib == "libcxx" || Tab.for_name("boost-python").stdlib == "libcxx")
+    odie "boost and boost-python must be built against libstdc++ on Mavericks. One way to achieve this, is to use GCC to compile both libraries."
+  end
+
```
Comment (dpo) [inline graph-tool.rb]: Sorry, this should be in `install`.

**Comment by dpo** [comment]
It failed correctly on 10.9 but it still fails due to the lazy symbol thing on 10.10. Might need to request clang >= 703.

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by dpo** [comment]
@count0 Sorry for the long delay. Do you want to update your PR to 2.16 and we'll try and wrap this up?

**Comment by kikohs** [comment]
Hey, what's the status of this formula ? Could it be possible to have the latest graph-tool? Thanks

**Comment by count0** [comment]
@dpo I've updated to version 2.16, but I'm still at loss on how to fix this. I simply can't hit the mark with the clang whitelisting/blacklisting that satisfies the test bot.

**Comment by dpo** [comment]
OSX 10.9 is tricky because there's a requirement that `boost` and `boost-python` be built with gcc. But there's also a build error on 10.10:
```
In file included from /usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/tuple:39:0,
                 from /usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/functional:55,
                 from /usr/local/include/boost/config/no_tr1/functional.hpp:21,
                 from /usr/local/include/boost/function/detail/prologue.hpp:14,
                 from /usr/local/include/boost/function/function_template.hpp:13,
                 from /usr/local/include/boost/function/detail/maybe_include.hpp:13,
                 from /usr/local/include/boost/function/function0.hpp:11,
                 from /usr/local/include/boost/python/errors.hpp:13,
                 from /usr/local/include/boost/python/handle.hpp:11,
                 from /usr/local/include/boost/python/args_fwd.hpp:10,
                 from /usr/local/include/boost/python/args.hpp:10,
                 from /usr/local/include/boost/python.hpp:11,
                 from ../../../../src/graph/community_old/graph_blockmodel_covariates.cc:19:
/usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/array: In constructor 'constexpr std::array<std::pair<long unsigned int, long unsigned int>, 2ul>::array()':
/usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/array:90:12:   in constexpr expansion of 'std::pair<long unsigned int, long unsigned int>()'
/usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/array:90:12: internal compiler error: in cxx_eval_call_expression, at cp/constexpr.c:1449
     struct array
            ^~~~~
```

**Comment by count0** [comment]
Internal compiler errors are, by definition, compiler bugs. Hence we need to see if clang will work for 10.10 instead, but I think we went that road before.

**Comment by dpo** [comment]
2.16 built for me on 10.10 with clang but not with gcc-6.1.

**Comment by count0** [comment]
@dpo and what clang build would that be?

**Comment by dpo** [comment]
I have build 700.1.81.

**Comment by count0** [comment]
@dpo We have a fail on mavericks only because of the libstdc++ and libc++ thing. Yosemite and  El Capitain seem to be working.

**Comment by dpo** [comment]
Working on it.

**Comment by dpo** [comment]
I forced compilation of `boost` and `boost-python` with libstdc++ but now I'm getting an internal compiler error on Mavericks:
```
/usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/array: In constructor 'constexpr std::array<std::pair<long unsigned int, long unsigned int>, 2ul>::      array()':
/usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/array:90:12:   in constexpr expansion of 'std::pair<long unsigned int, long unsigned int>()'
/usr/local/Cellar/gcc/6.1.0/include/c++/6.1.0/array:90:12: internal compiler error: in cxx_eval_call_expression, at cp/constexpr.c:1449
     struct array
            ^~~~~
```
It's not the first time I'm regretting the gcc 6.1 upgrade. I've encountered internal errors in gfortran too. I'm not sure why they don't show up on 10.10 and 10.11.

At this point, I think we just have to disable the bottle for Mavericks.

**Comment by count0** [comment]
I agree with just disabling it for Mavericks. Should I just remove the line from the formula?

**Comment by dpo** [comment]
Because there's no syntax to explicitly disable bottling for one architecture, let's add this to the formula:
```ruby
if MacOS.version == :mavericks
  fails_with :gcc => "6" do
    cause "Internal compiler error"
  end
end
```
which we'll hopefully be able to remove at the next gcc upgrade. `graph-tool` builds on 10.9 with gcc-5, so users will just have to `brew tap homebrew/versions`, `brew install gcc5` and use it to build `boost`, `boost-python` and `graph-tool`. Does that sound reasonable?

**Comment by count0** [comment]
@dpo Yes, that is reasonable. I've made the change.

**Comment by dpo** [comment]
It seems we're good to go. Would you mind squashing all your commits?

**Comment by count0** [comment]
Can't you do this automatically when doing merge? i.e. https://help.github.com/articles/about-pull-request-merge-squashing/

Otherwise I have to close this pull request, and do another one.

**Comment by dpo** [comment]
Nevermind, I'll do it. It doesn't happen automatically because there's more to just merging a Homebrew PR.

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by count0** [comment]
This is awesome, thanks.

**Pull request closed without merging**